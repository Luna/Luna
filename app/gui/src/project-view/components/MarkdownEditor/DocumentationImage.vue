<script setup lang="ts">
import {
  injectDocumentationImageUrlTransformer,
  type TransformUrlResult,
} from '@/components/MarkdownEditor/imageUrlTransformer'
import { computedAsync } from '@vueuse/core'
import { computed, onUnmounted, type Ref } from 'vue'
import { Ok } from 'ydoc-shared/util/data/result'

const DEFAULT_ALT_TEXT = 'Image'

const props = defineProps<{
  src: string
  alt: string
}>()

const urlTransformer = injectDocumentationImageUrlTransformer(true)

// NOTE: Garbage-collecting image data when the `src` changes is not implemented. Current users of `DocumentationImage`
// don't change the `src` after creating an image.
const data: Ref<TransformUrlResult | undefined> =
  urlTransformer ?
    computedAsync(() => urlTransformer.transformUrl(props.src), undefined, {
      onError: console.error,
    })
  : computed(() => Ok({ url: props.src }))

const title = computed(() =>
  data.value == null ? 'Loading'
  : !data.value.ok ? data.value.error.message()
  : props.alt !== DEFAULT_ALT_TEXT ? props.alt
  : '',
)

const alt = props.alt ? props.alt : DEFAULT_ALT_TEXT

onUnmounted(() => {
  if (data.value?.ok) data.value.value.dispose?.()
})
</script>

<template>
  <img
    :src="data?.ok ? data.value.url : ''"
    :alt="alt"
    :title="title"
    :class="{ uploading: data?.ok && data.value.uploading?.value }"
  />
</template>
