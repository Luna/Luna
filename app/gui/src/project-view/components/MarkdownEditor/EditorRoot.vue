<script setup lang="ts">
import { ref } from 'vue'

const rootElement = ref<HTMLElement>()
defineExpose({ rootElement })
</script>

<template>
  <div
    ref="rootElement"
    class="EditorRoot"
    @keydown.arrow-left.stop
    @keydown.arrow-right.stop
    @keydown.arrow-up.stop
    @keydown.arrow-down.stop
    @keydown.enter.stop
    @keydown.backspace.stop
    @keydown.delete.stop
    @wheel.stop.passive
    @contextmenu.stop
  ></div>
</template>

<style scoped>
.EditorRoot {
  width: 100%;
  height: 100%;
}

:deep(.cm-content) {
  font-family: var(--font-sans);
}

:deep(.cm-scroller) {
  /* Prevent touchpad back gesture, which can be triggered while panning. */
  overscroll-behavior: none;
}

.EditorRoot :deep(.cm-editor) {
  position: relative;
  width: 100%;
  height: 100%;
  opacity: 1;
  color: black;
  font-size: 12px;
  outline: none;
}
</style>
