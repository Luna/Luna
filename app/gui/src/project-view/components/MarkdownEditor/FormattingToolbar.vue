<script setup lang="ts">
import BlockTypeMenu from '@/components/MarkdownEditor/BlockTypeMenu.vue'
import FormatPropertiesBar from '@/components/MarkdownEditor/FormatPropertiesBar.vue'
import type { UseFormatting } from '@/components/MarkdownEditor/formatting'

const props = defineProps<{ formatting: UseFormatting }>()

const { blockType } = props.formatting
</script>

<template>
  <div class="FormattingToolbar">
    <div class="section">
      <BlockTypeMenu :modelValue="blockType.state.value" @update:modelValue="blockType.set" />
    </div>
    <div class="section">
      <FormatPropertiesBar :formatting="formatting" />
    </div>
  </div>
</template>

<style scoped>
.FormattingToolbar {
  --color-frame-bg: white;
  display: flex;
  height: 100%;
  flex-direction: row;
  align-items: center;
  gap: 6px;
  line-height: 16px;
}

.section {
  display: flex;
  background-color: white;
  border-radius: var(--radius-full);
  gap: 4px;
  padding: 4px;
}

.FormattingToolbar :deep(.iconLabel) {
  margin-left: 4px;
  padding-right: 4px;
}
</style>
