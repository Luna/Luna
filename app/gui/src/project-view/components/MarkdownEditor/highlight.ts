import { syntaxHighlighting } from '@codemirror/language'
import type { Extension } from '@codemirror/state'
import { type Tag, tagHighlighter, tags } from '@lezer/highlight'

const tagNames: (keyof typeof tags)[] = [
  'comment',
  'lineComment',
  'blockComment',
  'docComment',
  'name',
  'variableName',
  'typeName',
  'tagName',
  'propertyName',
  'attributeName',
  'className',
  'labelName',
  'namespace',
  'macroName',
  'literal',
  'string',
  'docString',
  'character',
  'attributeValue',
  'number',
  'integer',
  'float',
  'bool',
  'regexp',
  'escape',
  'color',
  'url',
  'keyword',
  'self',
  'null',
  'atom',
  'unit',
  'modifier',
  'operatorKeyword',
  'controlKeyword',
  'definitionKeyword',
  'moduleKeyword',
  'operator',
  'derefOperator',
  'arithmeticOperator',
  'logicOperator',
  'bitwiseOperator',
  'compareOperator',
  'updateOperator',
  'definitionOperator',
  'typeOperator',
  'controlOperator',
  'punctuation',
  'separator',
  'bracket',
  'angleBracket',
  'squareBracket',
  'paren',
  'brace',
  'content',
  'heading',
  'heading1',
  'heading2',
  'heading3',
  'heading4',
  'heading5',
  'heading6',
  'contentSeparator',
  'list',
  'quote',
  'emphasis',
  'strong',
  'link',
  'monospace',
  'strikethrough',
  'inserted',
  'deleted',
  'changed',
  'invalid',
  'meta',
  'documentMeta',
  'annotation',
  'processingInstruction',
]

/**
 * Defines an {@link Extension} that applies a highlighting CSS class for any {@link Tag} with a provided class mapping.
 * @param css A mapping from {@link Tag} names to CSS class names.
 */
export function highlightStyle(css: Record<string, string>): Extension {
  const modTagClasses = (mod: keyof typeof tags) =>
    tagNames.map((tag) => ({
      tag: (tags[mod] as any)(tags[tag]) as Tag,
      class: `${tags[mod]}-${tag}`,
    }))
  const tagClasses = tagNames.map((tag) => ({ tag: tags[tag] as Tag, class: css[tag] ?? tag }))
  return syntaxHighlighting(
    tagHighlighter([
      ...tagClasses,
      ...modTagClasses('definition'),
      ...modTagClasses('constant'),
      ...modTagClasses('function'),
      ...modTagClasses('standard'),
      ...modTagClasses('local'),
      ...modTagClasses('special'),
    ]),
  )
}
