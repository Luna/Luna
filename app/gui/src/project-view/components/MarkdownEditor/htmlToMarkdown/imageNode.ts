import { type TextMatchTransformer } from '@lexical/markdown'
import {
  type DOMConversionMap,
  type DOMConversionOutput,
  type NodeKey,
  $applyNodeReplacement,
  DecoratorNode,
} from 'lexical'

interface ImagePayload {
  altText: string
  key?: NodeKey | undefined
  src: string
}

function $convertImageElement(domNode: Node): null | DOMConversionOutput {
  if (domNode instanceof HTMLImageElement) {
    const { alt: altText, src } = domNode
    const node = $createImageNode({ altText, src })
    return { node }
  }
  return null
}

function notImplemented(): never {
  throw new Error('not implemented')
}

/** TODO: Add docs */
export class ImageNode extends DecoratorNode<void> {
  __src: string
  __altText: string

  /** TODO: Add docs */
  static override getType(): string {
    return 'image'
  }

  /** TODO: Add docs */
  static override clone(node: ImageNode): ImageNode {
    return new ImageNode(node.__src, node.__altText, node.__key)
  }

  /** TODO: Add docs */
  static override importDOM(): DOMConversionMap | null {
    return {
      img: (_node: Node) => ({
        conversion: $convertImageElement,
        priority: 0,
      }),
    }
  }

  /** TODO: Add docs */
  constructor(src: string, altText: string, key?: NodeKey) {
    super(key)
    this.__src = src
    this.__altText = altText
  }

  /** TODO: Add docs */
  override exportJSON() {
    return {
      altText: this.getAltText(),
      src: this.getSrc(),
      type: 'image',
      version: 1,
    }
  }

  /** TODO: Add docs */
  getSrc(): string {
    return this.__src
  }

  /** TODO: Add docs */
  getAltText(): string {
    return this.__altText
  }

  /** Not used, but required by API */
  override decorate() {}

  static override importJSON = notImplemented
  override exportDOM = notImplemented
  override createDOM = notImplemented
  override updateDOM = notImplemented
}

/** Type predicate for {@link ImageNode} */
export function $isImageNode(node: unknown): node is ImageNode {
  return node instanceof ImageNode
}

function $createImageNode({ altText, src, key }: ImagePayload): ImageNode {
  return $applyNodeReplacement(new ImageNode(src, altText, key))
}

export const IMAGE: TextMatchTransformer = {
  dependencies: [ImageNode],
  export: (node) => {
    if (!$isImageNode(node)) return null
    return `![${node.getAltText()}](${node.getSrc()})`
  },
  regExp: /$^/,
  type: 'text-match',
}
