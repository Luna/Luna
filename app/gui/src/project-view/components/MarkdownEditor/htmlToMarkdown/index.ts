/**
 * Convert the given HTML text to a Markdown approximation. The supporting libraries will be loaded the first time this
 * is called.
 */
export async function htmlToMarkdown(html: string): Promise<string> {
  const { htmlToMarkdownImpl } = await import(
    '@/components/MarkdownEditor/htmlToMarkdown/htmlToMarkdownImpl'
  )
  return htmlToMarkdownImpl(html)
}
