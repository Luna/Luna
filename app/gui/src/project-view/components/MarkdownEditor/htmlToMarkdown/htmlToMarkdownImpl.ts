import { IMAGE } from '@/components/MarkdownEditor/htmlToMarkdown/imageNode'
import { LINK } from '@/components/MarkdownEditor/htmlToMarkdown/linkNode'
import { $generateNodesFromDOM } from '@lexical/html'
import { $convertToMarkdownString, TRANSFORMERS } from '@lexical/markdown'
import { $insertNodes, createEditor } from 'lexical'

const domParser = new DOMParser()

const transformers = [IMAGE, LINK, ...TRANSFORMERS]

/** Convert the given HTML text to a Markdown approximation. */
export function htmlToMarkdownImpl(html: string): string {
  const dom = domParser.parseFromString(html, 'text/html')
  const editor = createEditor({
    nodes: transformers.flatMap((transformer) =>
      'dependencies' in transformer ? transformer.dependencies : [],
    ),
  })
  editor.update(() => $insertNodes($generateNodesFromDOM(editor, dom)))
  return editor.read(() => $convertToMarkdownString(transformers))
}
