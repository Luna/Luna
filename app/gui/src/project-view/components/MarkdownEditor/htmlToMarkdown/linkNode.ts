import { $isImageNode, IMAGE } from '@/components/MarkdownEditor/htmlToMarkdown/imageNode'
import { $isLinkNode, LinkNode } from '@lexical/link'
import { type Transformer } from '@lexical/markdown'
import { $isTextNode } from 'lexical'
export { LinkNode } from '@lexical/link'

export const LINK: Transformer = {
  dependencies: [LinkNode],
  export: (node, exportChildren, exportFormat) => {
    if (!$isLinkNode(node)) return null
    const title = node.getTitle()
    const linkContent =
      title ?
        `[${node.getTextContent()}](${node.getURL()} "${title}")`
      : `[${node.getTextContent()}](${node.getURL()})`
    const firstChild = node.getFirstChild()
    // Add text styles only if link has single text node inside. If it's more
    // than one we ignore it as markdown does not support nested styles for links
    if (node.getChildrenSize() === 1 && $isTextNode(firstChild)) {
      return exportFormat(firstChild, linkContent)
    } else if (node.getChildrenSize() === 1 && $isImageNode(firstChild)) {
      // Images sometimes happen to be inside links (when importing nodes from HTML).
      // The link is not important for us (this type of layout is not supported in markdown),
      // but we want to display the image.
      return IMAGE.export!(firstChild, exportChildren, exportFormat)
    } else {
      return linkContent
    }
  },
  regExp: /$^/,
  type: 'text-match',
}
