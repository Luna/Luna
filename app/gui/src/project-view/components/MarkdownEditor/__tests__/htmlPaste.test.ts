import { htmlToMarkdown } from '@/components/MarkdownEditor/htmlToMarkdown'
import * as fs from 'node:fs'
import { expect, test } from 'vitest'

test.each(['google_docs1.html.bin', 'preppin_data_2024w1.html.bin'])(
  'htmlToMarkdown: %s',
  async (file) => {
    const html = fs.readFileSync(`${__dirname}/fixtures/${file}`).toString()
    expect(await htmlToMarkdown(html)).toMatchSnapshot()
  },
)
