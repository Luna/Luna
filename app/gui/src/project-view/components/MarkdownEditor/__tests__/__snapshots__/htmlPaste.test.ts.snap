// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`htmlToMarkdown: google_docs1.html.bin 1`] = `
"# Title

Subtitle

# Heading1

## Heading2

### Heading3

#### Heading4

##### Heading5

###### *Heading6*

**Bold**, *italic*, and *underlined*.

Unordered list:

- Item
- Another

Numbered list:

1. First
2. Second

Link: [Text](http://example.com)

![](https://lh7-rt.googleusercontent.com/docsz/AD_4nXeCL-N-bkbZ3JKJqp86wjPGiXe0hSlYQVnr6nqfI4k37Mrf0USAOhbXyIyXzdIVhuD60vPN4BHHWf64F5b80nonbX63keL0BJRe2KsBQMInXS_Sn6Elj-SdB34CUc3d1xOllHf7WA?key=oYRu0j12MJwt-XgjX8Oamy3V)"
`;

exports[`htmlToMarkdown: preppin_data_2024w1.html.bin 1`] = `
"### 2024: Week 1 - Prep Air's Flow Card

[January 03, 2024](https://preppindata.blogspot.com/2024/01/2024-week-1-prep-airs-flow-card.html "permanent link")

Created by: Carl Allchin

Welcome to a New Year of Preppin' Data challenges. For anyone new to the challenges then let us give you an overview how the weekly challenge works. 

- Each Wednesday the Preppin' crew (Jenny, myself or a guest contributor) drop a data set(s) that requires some reshaping and/or cleaning to get it ready for analysis. 
    - You can use any tool or language you want to do the reshaping (we build the challenges in Tableau Prep but love seeing different tools being learnt / tried).
    - Share your solution on LinkedIn, Twitter/X, GitHub or the [Tableau Forums](https://community.tableau.com/s/group/0F94T000000gQqoSAE/preppindata)
    - Fill out [our tracker](https://docs.google.com/forms/d/e/1FAIpQLSdZzudRXwUvjhWwNawwz1kGXcYeQ9gBnAhCOvlA7qEDknGu4A/viewform) so you can monitor your progress and involvement
- The following Tuesday we will post a written solution in Tableau Prep (thanks Tom) and a video walkthrough too (thanks Jenny)

As with each January for the last few years, we'll set a number of challenges aimed at beginners. This is a great way to learn a number of fundamental data preparation skills or a chance to learn a new tool — New Year's resolution anyone? So on to this year's challenges!

### Context

At Preppin' Data we use a number of (mock) companies to look at the challenges they have with their data. For January, we're going to focus on our own airline, Prep Air. The airline has introduced a new loyalty card called the Flow Card. We need to clean up a number of data sets to determine how well the card is doing. 

The first task is setting some context for later weeks by understanding how popular the Flow Card is. Our stakeholder would like two data sets about our passengers. One data set for card users and one data set for those who don't use the card. 

### Input

There is one input file. You can [download it from here](https://drive.google.com/file/d/1STVYZvXzfGMuEq9Yq3yYOmCDCFq4iB0Z/view?usp=share_link).

![](https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEhyWUGjqNfzjbejKFop5EhJ5AvmFduIULRfKCy8CVzOyaqJQHQhyphenhyphen6Vgna7P_Ozfl0WvIZgd-JhihJVQJWRbkSeu9BbI8bjJ2_bFJg-OaOA8VE9yV8UWaFYL1xiAxGKGwIyGW9ebPqsy31xiw7SCQ610Y__4KiqoZdbhpqZssBrmbFLXpQUY_7nfbE1gHN9v/w640-h160/Screenshot%202023-12-12%20at%2020.37.44.png)

### Requirements

- Input the data
- Split the Flight Details field to form:
    - Date 
    - Flight Number
    - From
    - To
    - Class
    - Price
- Convert the following data fields to the correct data types:
    - Date to a date format
    - Price to a decimal value
- Change the Flow Card field to Yes / No values instead of 1 / 0
- Create two tables, one for Flow Card holders and one for non-Flow Card holders
- Output the data sets

### Output

Both outputs have the same data structure:

9 data fields:

- Date
- Flight Number
- From
- To
- Class
- Price
- Flow Card?
- Bags Checked
- Meal Type

Output 1

![](https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEjZvuMPD03wpttqY9OsQPWCC0MpBio6CDLUEfN_D9UW25P0vAYqTHITYkuuHxu5mSNjBnCpjg8iksGvW0u8ee2JOcfjxocZ6ob2CwMDtQw18BdYCv9ZJwDwdKXi2M6uMvv5b3jEYB2egkI1D7QEMVVn7xPBE9dZLizrFRx30g7bblg0f6e4OFjM1uGZAxEm/w640-h118/Screenshot%202023-12-12%20at%2021.09.34.png)

Rows: 1883 (1884 including the header)


Output 2

![](https://blogger.googleusercontent.com/img/b/R29vZ2xl/AVvXsEi1WTdeAiUOCwQncp073UYOxBTo_LCbL8L_AiEOr_BBzfkRCkiboiK9r5pF2lMCCN5p9iddBCgbpz5bsvgod2W9jmBLZepuk6uWOQaFMz6mhKTVLtgxfVdjWWM8E2BFGeNzALsz81YZ4FdswF4_KshBS2JHsXUVIEt1qlS4obFDTnA368aVxLORbCuqsvjG/w640-h102/Screenshot%202023-12-12%20at%2021.10.25.png)




Rows: 1895 (1896 including the header)

You can download the [output from here](https://drive.google.com/drive/folders/1-lt4XWZSn92gCBUiMT6ZDAoEzGyUZGJm?usp=share_link).




After you finish the challenge make sure to fill in the [participation tracker](https://docs.google.com/forms/d/e/1FAIpQLSdZzudRXwUvjhWwNawwz1kGXcYeQ9gBnAhCOvlA7qEDknGu4A/viewform), then share your solution on Twitter using [#PreppinData](https://twitter.com/search?q=%23preppindata&src=typed_query) and tagging [@Datajedininja](https://twitter.com/Datajedininja), [@JennyMartinDS14](https://twitter.com/JennyMartinDS14) & [@TomProwse1](https://twitter.com/TomProwse1)



You can also post your solution on the [Tableau Forum](https://community.tableau.com/s/group/0F94T000000gQqoSAE/preppindata) where we have a Preppin' Data community page. Post your solutions and ask questions if you need any help! "
`;
